core/parentchain/indirect-calls-executor/src/indirect_calls_executor.rs

```diff
@@ -48,53 +48,68 @@ pub trait ExecuteIndirectCalls {
 		ParentchainBlock: ParentchainBlockTrait<Hash = H256>;
 }
 
-pub struct IndirectCallsExecutor<ShieldingKeyRepository, StfExecutor, TopPoolAuthor> {
+pub struct IndirectCallsExecutor<ShieldingKeyRepository, StfEnclaveSigner, TopPoolAuthor> {
 	shielding_key_repo: Arc<ShieldingKeyRepository>,
-	stf_executor: Arc<StfExecutor>,
+	stf_enclave_signer: Arc<StfEnclaveSigner>,
 	top_pool_author: Arc<TopPoolAuthor>,
 }
 
-impl<ShieldingKeyRepository, StfExecutor, TopPoolAuthor>
-	IndirectCallsExecutor<ShieldingKeyRepository, StfExecutor, TopPoolAuthor>
+impl<ShieldingKeyRepository, StfEnclaveSigner, TopPoolAuthor>
+	IndirectCallsExecutor<ShieldingKeyRepository, StfEnclaveSigner, TopPoolAuthor>
 where
 	ShieldingKeyRepository: AccessKey,
-	<ShieldingKeyRepository as AccessKey>::KeyType:
-		ShieldingCryptoDecrypt<Error = itp_sgx_crypto::Error>,
-	StfExecutor: StfExecuteShieldFunds,
+	<ShieldingKeyRepository as AccessKey>::KeyType: ShieldingCryptoDecrypt<Error = itp_sgx_crypto::Error>
+		+ ShieldingCryptoEncrypt<Error = itp_sgx_crypto::Error>,
+	StfEnclaveSigner: StfEnclaveSigning,
 	TopPoolAuthor: AuthorApi<H256, H256> + Send + Sync + 'static,
 {
 	pub fn new(
 		shielding_key_repo: Arc<ShieldingKeyRepository>,
-		stf_executor: Arc<StfExecutor>,
+		stf_enclave_signer: Arc<StfEnclaveSigner>,
 		top_pool_author: Arc<TopPoolAuthor>,
 	) -> Self {
-		IndirectCallsExecutor { shielding_key_repo, stf_executor, top_pool_author }
+		IndirectCallsExecutor { shielding_key_repo, stf_enclave_signer, top_pool_author }
 	}
 
-	fn handle_shield_funds_xt(&self, xt: &UncheckedExtrinsicV4<ShieldFundsFn>) -> Result<()> {
-		let (call, account_encrypted, amount, shard) = &xt.function;
+	fn handle_shield_funds_xt(&self, xt: UncheckedExtrinsicV4<ShieldFundsFn>) -> Result<()> {
+		let (call, account_encrypted, amount, shard) = xt.function;
 		info!("Found ShieldFunds extrinsic in block: \nCall: {:?} \nAccount Encrypted {:?} \nAmount: {} \nShard: {}",
         	call, account_encrypted, amount, bs58::encode(shard.encode()).into_string());
 
 		debug!("decrypt the account id");
 
 		let shielding_key = self.shielding_key_repo.retrieve_key()?;
-		let account_vec = shielding_key.decrypt(account_encrypted)?;
+		let account_vec = shielding_key.decrypt(&account_encrypted)?;
 
 		let account = AccountId::decode(&mut account_vec.as_slice())?;
 
-		self.stf_executor.execute_shield_funds(account, *amount, shard)?;
+		let enclave_account_id = self.stf_enclave_signer.get_enclave_account();
+		let trusted_call = TrustedCall::balance_shield(enclave_account_id, account, amount);
+		let signed_trusted_call =
+			self.stf_enclave_signer.sign_call_with_self(&trusted_call, &shard)?;
+		let trusted_operation = TrustedOperation::indirect_call(signed_trusted_call);
+
+		let encrypted_trusted_call = shielding_key.encrypt(&trusted_operation.encode())?;
+		self.submit_trusted_call(shard, encrypted_trusted_call);
```

Comment:
This is the new handling of shield funds. We get the parameters from the extrinsic and create a signed trusted call, which then is submitted to the TOP pool.